File: typescript/ts_sdk/src/wns/binding.ts
import { resolveDidWbaDocument } from '../authentication/did-wba.js';
import type { DidDocument } from '../authentication/types.js';
import { HandleBindingError } from '../errors/index.js';
import type {
  BindingVerificationOptions,
  BindingVerificationResult,
  HandleServiceEntry,
} from './types.js';
import { HandleStatus } from './types.js';
import { resolveHandle } from './resolver.js';
import { buildResolutionUrl, validateHandle } from './validator.js';
import { ANP_HANDLE_SERVICE_TYPE as ANP_HANDLE_SERVICE_TYPE_VALUE } from './types.js';

export async function verifyHandleBinding(
  handle: string,
  options: BindingVerificationOptions = {}
): Promise<BindingVerificationResult> {
  const bareHandle = handle.startsWith('wba://') ? handle.slice('wba://'.length) : handle;
  let localPart: string;
  let domain: string;
  try {
    [localPart, domain] = validateHandle(bareHandle);
  } catch (error) {
    return {
      isValid: false,
      handle: bareHandle,
      did: '',
      forwardVerified: false,
      reverseVerified: false,
      errorMessage: (error as Error).message,
    };
  }

  const normalizedHandle = `${localPart}.${domain}`;

  try {
    const resolution = await resolveHandle(normalizedHandle, options.resolutionOptions);
    if (resolution.status !== HandleStatus.Active) {
      return {
        isValid: false,
        handle: normalizedHandle,
        did: resolution.did,
        forwardVerified: false,
        reverseVerified: false,
        errorMessage: `Handle status is '${resolution.status}', expected 'active'`,
      };
    }

    if (!resolution.did.startsWith('did:wba:')) {
      return {
        isValid: false,
        handle: normalizedHandle,
        did: resolution.did,
        forwardVerified: true,
        reverseVerified: false,
        errorMessage: 'DID does not use did:wba method',
      };
    }

    const didDomain = resolution.did.split(':')[2] ?? '';
    if (didDomain.toLowerCase() !== domain) {
      return {
        isValid: false,
        handle: normalizedHandle,
        did: resolution.did,
        forwardVerified: true,
        reverseVerified: false,
        errorMessage: `Domain mismatch: handle domain '${domain}' != DID domain '${didDomain}'`,
      };
    }

    const didDocument =
      options.didDocument ??
      (await resolveDidWbaDocument(resolution.did, false, options.didResolutionOptions));
    const handleServices = extractHandleServiceFromDidDocument(didDocument);
    const reverseVerified = handleServices.some(
      (service) => matchesHandleServiceDomain(service.serviceEndpoint, domain)
    );
    if (!reverseVerified) {
      return {
        isValid: false,
        handle: normalizedHandle,
        did: resolution.did,
        forwardVerified: true,
        reverseVerified: false,
        errorMessage:
          `DID Document does not contain an ${ANP_HANDLE_SERVICE_TYPE_VALUE} entry ` +
          `whose HTTPS domain matches '${domain}'`,
      };
    }

    return {
      isValid: true,
      handle: normalizedHandle,
      did: resolution.did,
      forwardVerified: true,
      reverseVerified: true,
    };
  } catch (error) {
    if (error instanceof HandleBindingError) {
      throw error;
    }
    return {
      isValid: false,
      handle: normalizedHandle,
      did: '',
      forwardVerified: false,
      reverseVerified: false,
      errorMessage: (error as Error).message,
    };
  }
}

export function buildHandleServiceEntry(
  did: string,
  localPart: string,
  domain: string
): HandleServiceEntry {
  return {
    id: `${did}#handle`,
    type: ANP_HANDLE_SERVICE_TYPE_VALUE,
    serviceEndpoint: buildResolutionUrl(localPart, domain),
  };
}

export function extractHandleServiceFromDidDocument(didDocument: DidDocument): HandleServiceEntry[] {
  return (didDocument.service ?? [])
    .filter((service) => service.type === ANP_HANDLE_SERVICE_TYPE_VALUE)
    .map((service) => ({
      id: String(service.id),
      type: String(service.type),
      serviceEndpoint: String(service.serviceEndpoint),
    }));
}

function matchesHandleServiceDomain(serviceEndpoint: string, expectedDomain: string): boolean {
  try {
    const parsed = new URL(serviceEndpoint);
    return (
      parsed.protocol === 'https:' && parsed.hostname.toLowerCase() === expectedDomain.toLowerCase()
    );
  } catch {
    return false;
  }
}

File: golang/wns/binding.go
package wns

import (
	"context"
	"fmt"
	"net/url"
	"strings"

	"github.com/agent-network-protocol/anp/golang/authentication"
)

// VerifyHandleBinding verifies forward and reverse handle binding.
func VerifyHandleBinding(ctx context.Context, handle string) BindingVerificationResult {
	return VerifyHandleBindingWithOptions(ctx, handle, BindingVerificationOptions{})
}

// VerifyHandleBindingWithOptions verifies forward and reverse handle binding with options.
func VerifyHandleBindingWithOptions(ctx context.Context, handle string, options BindingVerificationOptions) BindingVerificationResult {
	bareHandle := strings.TrimPrefix(handle, "wba://")
	localPart, domain, err := ValidateHandle(bareHandle)
	if err != nil {
		return BindingVerificationResult{IsValid: false, Handle: bareHandle, ErrorMessage: err.Error()}
	}
	normalized := localPart + "." + domain
	resolution, err := ResolveHandleWithOptions(ctx, normalized, options.ResolutionOptions)
	if err != nil {
		return BindingVerificationResult{IsValid: false, Handle: normalized, ErrorMessage: "forward resolution failed: " + err.Error()}
	}
	if resolution.Status != HandleStatusActive {
		return BindingVerificationResult{IsValid: false, Handle: normalized, DID: resolution.DID, ErrorMessage: fmt.Sprintf("handle status is '%s', expected 'active'", resolution.Status)}
	}
	if !strings.HasPrefix(resolution.DID, "did:wba:") {
		return BindingVerificationResult{IsValid: false, Handle: normalized, DID: resolution.DID, ForwardVerified: true, ErrorMessage: "DID does not use did:wba method"}
	}
	parts := strings.Split(resolution.DID, ":")
	didDomain := ""
	if len(parts) > 2 {
		didDomain = parts[2]
	}
	if strings.ToLower(didDomain) != domain {
		return BindingVerificationResult{IsValid: false, Handle: normalized, DID: resolution.DID, ForwardVerified: true, ErrorMessage: fmt.Sprintf("domain mismatch: handle domain '%s' != DID domain '%s'", domain, didDomain)}
	}
	didDocument := options.DidDocument
	if didDocument == nil {
		didDocument, err = authentication.ResolveDidWBADocumentWithOptions(ctx, resolution.DID, false, options.DidResolutionOptions)
		if err != nil {
			return BindingVerificationResult{IsValid: false, Handle: normalized, DID: resolution.DID, ForwardVerified: true, ErrorMessage: "failed to resolve DID document: " + err.Error()}
		}
	}
	handleServices := ExtractHandleServiceFromDIDDocument(didDocument)
	reverseVerified := false
	for _, service := range handleServices {
		if matchesHandleServiceDomain(service.ServiceEndpoint, domain) {
			reverseVerified = true
			break
		}
	}
	if !reverseVerified {
		return BindingVerificationResult{IsValid: false, Handle: normalized, DID: resolution.DID, ForwardVerified: true, ErrorMessage: fmt.Sprintf("DID Document does not contain an %s entry whose HTTPS domain matches '%s'", ANPHandleServiceType, domain)}
	}
	return BindingVerificationResult{IsValid: true, Handle: normalized, DID: resolution.DID, ForwardVerified: true, ReverseVerified: true}
}

// BuildHandleServiceEntry builds the reverse binding service entry.
func BuildHandleServiceEntry(did string, localPart string, domain string) HandleServiceEntry {
	return HandleServiceEntry{ID: did + "#handle", Type: ANPHandleServiceType, ServiceEndpoint: BuildResolutionURL(localPart, domain)}
}

// ExtractHandleServiceFromDIDDocument extracts ANPHandleService entries from a DID document.
func ExtractHandleServiceFromDIDDocument(didDocument map[string]any) []HandleServiceEntry {
	services, ok := didDocument["service"].([]any)
	if !ok {
		return nil
	}
	result := []HandleServiceEntry{}
	for _, entry := range services {
		service, ok := entry.(map[string]any)
		if ok && fmt.Sprintf("%v", service["type"]) == ANPHandleServiceType {
			result = append(result, HandleServiceEntry{ID: fmt.Sprintf("%v", service["id"]), Type: fmt.Sprintf("%v", service["type"]), ServiceEndpoint: fmt.Sprintf("%v", service["serviceEndpoint"])})
		}
	}
	return result
}

func matchesHandleServiceDomain(serviceEndpoint string, expectedDomain string) bool {
	parsed, err := url.Parse(serviceEndpoint)
	if err != nil {
		return false
	}
	return strings.EqualFold(parsed.Scheme, "https") && strings.EqualFold(parsed.Hostname(), expectedDomain)
}
